Support 9-byte command protocol for 0x8125

claudobahn opened the pull request:
Resolves #44 

Based on my collaboration w/ @gerth2 (see his PR #42) I've implemented support for 0x8125 controllers that we both have. There's inspirations from his PR to be found in mine (especially the command-line stuff), as well as a simplification of the protocol handling.

@mjg59 please have a look and let me know if I'm on the right track. I was trying to keep this pull request clean, yet still clean up some logic.

I've tested it on my controller and it's behaving as expected so far.

My type of controller looks like this: https://www.amazon.com/gp/product/B01DY56N8U

claudobahn commented on flux_led/__main__.py:
```diff
@@ -940,7 +940,30 @@ def setRgbw(self, r=None, g=None, b=None, w=None, persist=True,
+
+            # write mask, default to writing color and whites simultaneously
+            write_mask = 0x00
+            if self.rgbwprotocol:
```
If I remove this and always let it calculate the write_mask, I think that fixes the remaining test failures. The question is what's the right thing to do. Hoping @mjg59 can provide more insight into what the correct thing to do here is. I can confirm it works for my controller, but want to make sure it still works for others.

mjg59 commented on flux_led/__main__.py:
```diff
@@ -940,7 +947,29 @@ def setRgbw(self, r=None, g=None, b=None, w=None, persist=True,
+                # LEDENET devices support two white outputs for cold and warm. We set
+                # the second one here - if we're only setting a single white value,
+                # we set the second output to be the same as the first
+                if w2 is not None:
```
I think this will break the devices that previously set rgbwprotocol - they need this byte to be 0.

claudobahn replied:
I was doing code archeology, but without having some other data captures it's a little bit difficult. Were you able to dig up any other data captures?

My impression is there's 3 protocols currently:
LEDENET_ORIGINAL (5 bytes)
unnamed (8-byte) -- 4 channel (RGB + W)
LEDENET (9-byte) -- 5 channel (RGB + W + W)

Both the 8 and 9 byte variants use a write mask, the difference is how many white channels they send.

Currently only 0x25, 0x27, and 0x35 will send the second white channel.

Please go back and check out the diff for https://github.com/Danielhiversen/flux_led/commit/fc169a33a824c18a6054923963cb941c1a164f49 if you look at the before version, you'll see what I mean, I think.

mjg59 replied:
https://github.com/Danielhiversen/flux_led/blob/47c6e02e178dae19c871c9e1bde26ae035cd5213/flux_led/__main__.py is the commit just before the LEDENET_ORIGINAL was added. If you look at how rgbwprotocol is used, you'll see that they're 8-byte protocols that don't use the write mask. This only applies to types 0x04, 0x33 and 0x81.

claudobahn replied:
Oh, I should also add.. This is basically a mask for the controller. It allows you to set just the whites w/o having to send the colors. That's what the app does on its individual screens, it leaves the colors alone while changing the white, by masking out the colors.

If you send 0x00, then everything is updated, so if you don't include the colors you want set, then it'll overwrite them w/ 0, basically turning off the RGB LEDs. Don't think that would be intended..

mjg59 replied:
0x81 devices get an app screen that *does* allow RGB and W to be modified simultaneously, and the full RGBW value is always sent with the mask byte set to 0.

claudobahn replied:
Do you happen to have one of those? If so, could you try it out to see if the mask still works? If it does, that would be a functionality improvement, I think. If not (can't test or doesn't work), I'll revert that part.

claudobahn replied:
Ok, I've restored that behavior. Though I still wonder if those devices could support the write mask (and be better off with it). Also, since 0x33 is in the rgbwprotocol list, am I assuming correctly that it's rgbwcapable? I added it to that list, let me know if that's wrong.

claudobahn commented on flux_led/__main__.py:
```diff
@@ -877,14 +891,19 @@ def setRgbw(self, r=None, g=None, b=None, w=None, persist=True,
-        if self.badrgbw == True and (r != None and (w != None or w2 != None)):
-            self.setRgbw(w=w, w2=w2, retry=retry, persist=persist)
-            self.setRgbw(r, g, b, retry=retry, persist=persist)
-            return
```
The app only sends independent requests, that's correct. The app doesn't have a screen where both whites and colors can be set at the same time.

However if I use the flux_led command-line, I can set all 5 channels simultaneously. Confirmed that with my controller (0x25).

mjg59 replied:
Cool! Let's just do that, then.

Danielhiversen commented:
Is this ready for review and testing now?
What is the difference between this pr and https://github.com/Danielhiversen/flux_led/pull/42 ?

SmaginPV commented:
I also had the same issue, and according to the discussion, this PR will help me too. Let's merge it?

claudobahn commented:
@Danielhiversen yes, I've been using it in my HA install for a few weeks. I was mostly hoping for @mjg59 to confirm that my interpretation of how other devices deal with the protocol is correct.

Danielhiversen commented on flux_led/__main__.py:
```diff
@@ -59,7 +59,7 @@ def color_object_to_tuple(color):
 
         # see if it's already a color tuple
-        if type(color) is tuple and (len(color) == 3 or len(color == 4)):
+        if type(color) is tuple and (len(color) >= 3 and len(color) <= 5):
```
No need for parenthesis around `len(color) >= 3 and len(color) <= 5`

Danielhiversen replied:
Or change it to `len(color) in [3, 4, 5]`

Danielhiversen commented on flux_led/__main__.py:
```diff
@@ -87,7 +87,7 @@ def color_object_to_tuple(color):
         try:
             val = ast.literal_eval(color)
-            if type(val) is not tuple or (len(val) != 3 and len(val) != 4):
+            if type(val) is not tuple or len(val) < 3 or len(val) > 5:
```
`len(color) not  in [3, 4, 5]` might be more readable

claudobahn commented:
@SmaginPV this is the type of device I have: https://www.amazon.com/gp/product/B01DY56N8U and it works well with this PR.

Danielhiversen commented:
And what is the difference between this pr and #42 ?

claudobahn commented:
@Danielhiversen the main difference between this PR and #42 is one of approach. They both are intending to solve the same problem. I tried to keep the diff as readable as possible and also added a basic unit test.

I believe I've addressed your change requests.

Danielhiversen commented:
@gerth2 is it ok for you that we merge this instead of #42 ?

gerth2 commented:
No problem! I can give it a shot tonight to double check functionality
against both of my units.

On Jun 7, 2017 1:50 AM, "Daniel Høyer Iversen" <notifications@github.com>
wrote:

> @gerth2 <https://github.com/gerth2> is it ok for you that we merge this
> instead of #42 <https://github.com/Danielhiversen/flux_led/pull/42> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Danielhiversen/flux_led/pull/46#issuecomment-306703455>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEXw7v64aADFkLAdLr49O97qUelgYoHhks5sBkhKgaJpZM4NhcjB>
> .
>
